text/0000-landmarks.md

```diff
@@ -0,0 +1,86 @@
+
+```
+LandmarkNode {
+  LandmarkObservation {
+    common::Time time;
+    int trajectory_id;
+    
+    transform::Rigid3d transform;
+    double translation_weight;
+    double rotation_weight;
+  };
+  std::vector<LandmarkObservation> landmark_observations;
+  transform::Rigid3d global_pose;
+};
+```
+Note, that the `global_pose` is a vector of optimization parameters. Before the first sparse pose optimization run we can store extrapolated pose here instead and use it for rendering. Another option would be not populating the `global_pose` at all and not rendering the respective landmark on the map before optimization.
+
+
+### Optimization
+Landmarks will enter sparse pose graph optimization as additional independent variables. It increases the dimension of the optimization by `number_of_landmarks * number_of_parameters_per_landmark`, which could be undesirable when there is a significant number of landmarks. 
+
+#### Cost function
+Landmark sensor data contains only the landmark position relative to the robot. In order to formulate a landmark-to-trajectory cost functional the global pose of the robot at the moment of the observation has to be known. Since a landmark observation can be made at the time between two trajectory nodes and, moreover, the global poses of the trajectory nodes are state variables as well, the translation and rotation of the landmark at the moment can be interpolated using the previous and the next node, i.e. taking a convex combination of the translation components and slerp of the quaternions. This logic is already implemented within `TransformInterpolationBuffer`. 
+
+Residuals depend on the parameter block `[c_previous, c_next, c_landmark]`, i.e. on the global poses of the previous and next trajectory nodes and the global pose of the landmark itself:
+
+![component diagram](0000-assets/cost.png "Cost function")
+
+where `w` is the weight vector; `c` denotes global poses and `h` relative ones. 
+
+There will be no loss function as for intra submap constraints. 
+
+#### Constraints
+The cost function for landmark-to-trajectory constraints depends on the previous and next trajectory nodes. They are not known at the moment when the sensor data comes, but we can find them prior to the `RunOptimization` call using `AddWorkItem()` for landmark observations that were not yet processed.
+
+There are two ways of storing the landmark constraints in `PoseGraph`:
+
+1.  as a separate `std::vector<LandmarkConstraint> landmark_constraints_`, which would bloat the `PoseGraph` even more. Since the next RFC will talk about adding landmark-to-landmark constraints, I would try to avoid this option.
+1. in the current `std::vector<Constraint> constraints_`, which would lead to expanding the `Constraint` struct and adding  a new value to `enum Tag { INTRA_SUBMAP, INTER_SUBMAP } tag;`
+
+I think the second option is a bit cleaner. Both are ugly though.
```

Comment:
Can you explain why this is needed? It seems the map of `LandmarkNodes` which contains the observation already has all the necessary data. Why duplicate?

Reply:
It contains all necessary data except IDs of the previous and next nodes. Another solution would be to add these IDs to LandmarkNode proto and then, when a Landmark is observed, add it to some queue of landmarks for which we do not know previous/next nodes yet.